File: Leetcode/October_LeetCoding_Challenge/Day16_Search_a_2D_Matrix/Day16_Search_a_2D_Matrix.cpp
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        if (matrix.size() == 0)
            return false;
        int rows = matrix.size();
        int columns = matrix[0].size();

        int left = 0;
        int right = rows * columns - 1;

        while (left <= right)
        {
            int midpoint = left + (right - left) / 2; // Remember this method to find middle elements
            int midpoint_element = matrix[midpoint / columns][midpoint % columns];
            if (midpoint_element == target)
                return true;
            if (midpoint_element > target)
                right = midpoint + 1;
            if (midpoint_element < target)
                left = midpoint - 1;
        }
    }
};

File: LeetCode/Monthly Challenges/October_LeetCoding_Challenge/Day16_Search_A_2D_Matrix/Day16_Search_A_2D_Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0 )
            return false;
        
        int row = matrix.length;
        int col = matrix[0].length;
        int i = 0; 
        int j = col - 1;
        while(i < row && j >= 0){
            if(target == matrix[i][j])
                return true;
            else if(target < matrix[i][j]){
                j--;    
            }else if(target > matrix[i][j]){
                i++;
            }
            
        }
        return false;
    }
}
